### PR TITLE
feat: `SetFormula.successiveVariableElimination`

### PR DESCRIPTION
**Pull request by JonathanStarup**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetSubstitution.scala [hunk 1]
```diff
@@ -143,6 +143,14 @@ case class SetSubstitution(m: Map[Int, SetFormula]) {
     }
   }
 
+  /**
+    * Adds the disjoint binding `[x -> f]` to `this`.
+    *
+    * OBS: The [[SetFormula.Var]] of `this`, `x`, and `f` must not overlap.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/set/SetSubstitution.scala:149]: This would be costly to check, right?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/set/SetSubstitution.scala:149]: Like with an assertion.

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala [hunk 2]
```diff
@@ -872,4 +872,146 @@ object SetFormula {
...
 
+  /**
+    * The Successive Variable Elimination algorithm.
+    *
+    * Returns the most-general unifier of the equation `f ~ empty` where `fvs` is the free
+    * variables in `f`. If there is no unifier then [[NoSolutionException]] is thrown.
+    *
+    * Eliminates variables one-by-one from the given list `fvs` recursively.
+    *
+    * If the recursively building formula is ever larger than `recSizeThreshold` then
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:883]: Grammar. At least it reads strange to me.

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala [hunk 3]
```diff
@@ -872,4 +872,146 @@ object SetFormula {
...
+    * [[ComplexException]] is thrown. If `recSizeThreshold` is non-positive then there is no
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:884]: Why do it like this? I would just prefer people can use Int.MaxValue or whatever

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala [hunk 4]
```diff
@@ -872,4 +872,146 @@ object SetFormula {
...
+    * checking.
+    */
+  def successiveVariableElimination(f: SetFormula, fvs: List[Int], recSizeThreshold: Int): SetSubstitution = fvs match {
+    case Nil =>
+      // `emptyEquivalent` now treats the remaining constants as variables and goes through all
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:889]: Can you rephrase. Its hard to understand what should happen here. (I know what should happen, but re-read your sentence).

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala [hunk 5]
```diff
@@ -872,4 +872,146 @@ object SetFormula {
...
+      // assignments of those.
+      if (emptyEquivalent(f)) SetSubstitution.empty
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:891]: why not call it `isEmpty?` or `isEmptyEquiv`?  or similar.

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala [hunk 6]
```diff
@@ -872,4 +872,146 @@ object SetFormula {
...
+      else throw NoSolutionException()
+
+    case x :: xs =>
+      val fEmpty = SetSubstitution.singleton(x, Empty)(f)
+      val fUniv = SetSubstitution.singleton(x, Univ)(f)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:896]: I think it would really be better to keep the short names used elsewhere. They match those used in the papers; so its possible to lookup and compare.

**Review by magnus-madsen — COMMENTED** [review]
Mostly complaints about documentation :)

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetSubstitution.scala (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/unification/set/SetSubstitution.scala:149]: yeah I didn't think it was worth it. And technically the mapping `x -> x` is the same as not having `x` so we should also account for that in the check

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/unification/set/SetSubstitution.scala:149]: But we could constrain it to its less general form, and check that all mappings x -> f have f.isground and that the keys are disjoint. Also expensive but less cumbersome and its enough for the sve use

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala (hunk 3 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:884]: that would mean that we always call `f.size` which is expensive. But you can still call it with `MaxValue` if you want to.

File: main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala (hunk 6 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:896]: hmm, i dont agree. but not enough to insist. ill change it back

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala:896]: This I insist on. If I ever take it to Jaco he will not be able to read it :)

**Review by magnus-madsen — APPROVED** [review]
